### DISCUSSION TITLE
Controller returns old fields value when updating it's name dynamically

### DISCUSSION THREAD

**Discussion by ggat** [react-hook-form/react-hook-form#11602] [Q&A] [answered]
When changing Controller name dynamically, the Controller triggers its input re-render where `field.name` reflects updated name but `field.values` is of the old field. If the parent component re-renders once again then `field.name` and `field.values` are in sync again. Apparently it take 1 render cycle for the value to refresh.

I got an [example proof on CodeSandbox](https://codesandbox.io/p/sandbox/use-react-form-dynamic-controller-name-5njkkd?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clth5ndht00063b6hfjdv59jw%2522%252C%2522sizes%2522%253A%255B73.65044247787611%252C26.349557522123888%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clth5ndht00023b6ho4n09v2n%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clth5ndht00033b6h7i8bykhn%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clth5ndht00053b6h6a2h78ph%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clth5ndht00023b6ho4n09v2n%2522%253A%257B%2522id%2522%253A%2522clth5ndht00023b6ho4n09v2n%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clth5ndht00053b6h6a2h78ph%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clth5ndht00043b6hfv3j7dkt%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clth5ndht00053b6h6a2h78ph%2522%252C%2522activeTabId%2522%253A%2522clth5ndht00043b6hfv3j7dkt%2522%257D%252C%2522clth5ndht00033b6h7i8bykhn%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clth5ndht00033b6h7i8bykhn%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

- Click on "Swap name", in the console you'll see the output "value for banana apple" (wrong)
- Click on "Trigger Render", now the output is "value for banana banana" (right)

Thoughts? Do you think this might be a bug or expected behavior?

**Comment by Moshyfawn** [comment, marked as answer]
See [this FAQ](https://react-hook-form.com/faqs#Whyisdefaultvaluenotchangingcorrectlywithternaryoperator) for an explanation.

```diff
  <Controller
+   key={currentName}
    control={methods.control}
    name={currentName}
    render={({ field: { value, name } }) => {
      console.log(`value for ${name}`, value);
    }}
  />
```

**Reply by ggat** [reply]
Thanks @Moshyfawn, that was my original solution too, but I was not entirely sure, now I'm confident.

**Reply by kpotter-m2** [reply]
1 year later and this is still the fix, thank you!

**Comment by Mario-Eis** [comment]
Is there any downside in generally passing the field name as a key for every field in the whole application?
